crates/iceberg-catalog/src/service/storage/s3.rs

```diff
@@ -572,49 +407,169 @@ impl S3Profile {
             StoragePermissions::Read => "\"s3:GetObject\"",
             StoragePermissions::ReadWrite => "\"s3:GetObject\", \"s3:PutObject\"",
             StoragePermissions::ReadWriteDelete => {
-                "\"s3:GetObject\", \"s3:PutObject\", \"s3:DeleteObject\", \"s3:ListBucket\""
+                "\"s3:GetObject\", \"s3:PutObject\", \"s3:DeleteObject\""
             }
         }
     }
 
     fn get_aws_policy_string(
-        table_location: &str,
+        table_location: &Location,
         storage_permissions: StoragePermissions,
-    ) -> String {
-        let resource = format!(
-            "arn:aws:s3:::{}/*",
-            table_location.trim_start_matches("s3://")
+    ) -> Result<String, TableConfigError> {
+        let table_location = S3Location::try_from(table_location.clone()).map_err(|e| {
+            TableConfigError::Misconfiguration(
+                format!("Location is no valid S3 location: {e}").to_string(),
+            )
+        })?;
+        let bucket_arn = format!(
+            "arn:aws:s3:::{}",
+            table_location.bucket_name().trim_end_matches('/')
         );
-        format!(
+        let key = table_location.key().join("/");
+        let key = format!("{key}/");
+
+        Ok(format!(
             r#"{{
         "Version": "2012-10-17",
         "Statement": [
             {{
-                "Sid": "{}",
+                "Sid": "TableAccess",
```

Comment:
I vaguely recall a uniqueness requirement on Sid, could be that it's not relevant for inline policies, are you aware of this?

Reply:
It only needs to be unique inside one policy.

Docs:
You can use the Sid value as a description for the policy statement. In services that let you specify an ID element, such as SQS and SNS, the Sid value is just a sub-ID of the policy document ID. In IAM, the Sid value must be unique within a JSON policy.